### PR TITLE
crimson/os/heartbeat: make Heartbeat::send_failures() safe

### PR DESCRIPTION
**Pull request by xxhdx1985126**
Signed-off-by: Xuehan Xu <xxhdx1985126@163.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by xxhdx1985126** [comment]
@tchaikov Hi, I've modified the code as we discussed with the only exception that I kept the seastar::future<> return of send_failures just as Yingxin did, just in case someone would want to wait on that event in the future. Please take a look, thanks:-)

File: src/crimson/osd/heartbeat.cc [hunk 1]
```diff
@@ -386,26 +394,27 @@ seastar::future<> Heartbeat::send_heartbeats()
...
-      return monc.send_message(failure_report);
-    });
+  std::vector<seastar::future<>> futures;
+  for (auto [osd, failed_since] : failure_queue) {
+    if (failure_pending.count(osd)) {
+      continue;
+    }
+
+    auto failed_for = chrono::duration_cast<chrono::seconds>(
+	clock::now() - failed_since).count();
```
Comment (tchaikov) [inline src/crimson/osd/heartbeat.cc]: nit, could move `clock::now()` out of the loop.

File: src/crimson/osd/heartbeat.cc [hunk 2]
```diff
@@ -386,26 +394,27 @@ seastar::future<> Heartbeat::send_heartbeats()
...
-      failure_pending.emplace(osd, failure_info_t{failed_since,
-                                                  osdmap->get_addrs(osd)});
-      return monc.send_message(failure_report);
-    });
+  std::vector<seastar::future<>> futures;
+  for (auto [osd, failed_since] : failure_queue) {
+    if (failure_pending.count(osd)) {
+      continue;
+    }
+
```
Comment (tchaikov) [inline src/crimson/osd/heartbeat.cc]: i'd suggest remove the vertical spacing.

**Comment by xxhdx1985126** [comment]
@tchaikov Done:-)

**Pull request merged**